In [1]:
import numpy as np
import pandas as pd
from sqlalchemy import create_engine
import boto3
import s3fs
import awscli
import timeit
from datetime import datetime
from pandas.api.types import is_string_dtype
from pandas.api.types import is_numeric_dtype

In [2]:
s3 = boto3.resource('s3')

for bucket in s3.buckets.all():
    print(bucket.name)

team3-final-bucket
wshih-bucket


In [3]:
my_bucket = s3.Bucket('team3-final-bucket')

In [5]:
client = boto3.client('s3')
path = 's3://team3-final-bucket/sql_db/'

In [6]:
objects_df = pd.read_csv(path + 'objects_clean.csv', dtype={'status':str, 'entity_id':str})
objects_df.head()

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (3,5,7,8,9,10,11,13,14,17,18) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Unnamed: 0 object_id entity_type parent_id                name  \
0           0       c:1     Company       NaN            Wetpaint   
1           1      c:10     Company       NaN             Flektor   
2           2     c:100     Company       NaN               There   
3           3   c:10000     Company       NaN             MYWEBBO   
4           4   c:10001     Company       NaN  THE Movie Streamer   

     category_code     status  founded_at closed_at country_code  ...  \
0              web  operating  2005-10-17       NaN          USA  ...   
1      games_video   acquired         NaN       NaN          USA  ...   
2      games_video   acquired         NaN       NaN          USA  ...   
3  network_hosting  operating  2008-07-26       NaN          NaN  ...   
4      games_video  operating  2008-07-26       NaN          NaN  ...   

  investment_rounds invested_companies first_funding_at last_funding_at  \
0                 0                  0       2005-10-01      2008-05-19   
1                 0                  0              NaN             NaN   
2                 0                  0              NaN             NaN   
3                 0                  0              NaN             NaN   
4                 0                  0              NaN             NaN   

  funding_rounds  funding_total_usd  first_milestone_at last_milestone_at  \
0              3         39750000.0          2010-09-05        2013-09-18   
1              0                0.0                 NaN               NaN   
2              0                0.0          2003-02-01        2011-09-23   
3              0                0.0                 NaN               NaN   
4              0                0.0                 NaN               NaN   

  milestones  relationships  
0          5             17  
1          0              6  
2          4             12  
3          0              0  
4          0              0  

[5 rows x 25 columns]

In [7]:
objects_df.dtypes

Unnamed: 0               int64
object_id               object
entity_type             object
parent_id               object
name                    object
category_code           object
status                  object
founded_at              object
closed_at               object
country_code            object
state_code              object
city                    object
region                  object
first_investment_at     object
last_investment_at      object
investment_rounds        int64
invested_companies       int64
first_funding_at        object
last_funding_at         object
funding_rounds           int64
funding_total_usd      float64
first_milestone_at      object
last_milestone_at       object
milestones               int64
relationships            int64
dtype: object

In [9]:
is_string_dtype(objects_df['status'])

True

In [10]:
is_numeric_dtype(objects_df['status'])

False

In [11]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):  # more options can be specified also
    print(objects_df.dtypes)

Unnamed: 0               int64
object_id               object
entity_type             object
parent_id               object
name                    object
category_code           object
status                  object
founded_at              object
closed_at               object
country_code            object
state_code              object
city                    object
region                  object
first_investment_at     object
last_investment_at      object
investment_rounds        int64
invested_companies       int64
first_funding_at        object
last_funding_at         object
funding_rounds           int64
funding_total_usd      float64
first_milestone_at      object
last_milestone_at       object
milestones               int64
relationships            int64
dtype: object


In [12]:
objects_df.head()

Unnamed: 0 object_id entity_type parent_id                name  \
0           0       c:1     Company       NaN            Wetpaint   
1           1      c:10     Company       NaN             Flektor   
2           2     c:100     Company       NaN               There   
3           3   c:10000     Company       NaN             MYWEBBO   
4           4   c:10001     Company       NaN  THE Movie Streamer   

     category_code     status  founded_at closed_at country_code  ...  \
0              web  operating  2005-10-17       NaN          USA  ...   
1      games_video   acquired         NaN       NaN          USA  ...   
2      games_video   acquired         NaN       NaN          USA  ...   
3  network_hosting  operating  2008-07-26       NaN          NaN  ...   
4      games_video  operating  2008-07-26       NaN          NaN  ...   

  investment_rounds invested_companies first_funding_at last_funding_at  \
0                 0                  0       2005-10-01      2008-05-19   
1                 0                  0              NaN             NaN   
2                 0                  0              NaN             NaN   
3                 0                  0              NaN             NaN   
4                 0                  0              NaN             NaN   

  funding_rounds  funding_total_usd  first_milestone_at last_milestone_at  \
0              3         39750000.0          2010-09-05        2013-09-18   
1              0                0.0                 NaN               NaN   
2              0                0.0          2003-02-01        2011-09-23   
3              0                0.0                 NaN               NaN   
4              0                0.0                 NaN               NaN   

  milestones  relationships  
0          5             17  
1          0              6  
2          4             12  
3          0              0  
4          0              0  

[5 rows x 25 columns]

##### Converting date columns into date format

In [13]:
# Converting strings to date type
objects_df['founded_at'] = pd.to_datetime(objects_df.founded_at, format='%Y-%m-%d')
objects_df['closed_at'] = pd.to_datetime(objects_df.closed_at, format='%Y-%m-%d')
objects_df['first_investment_at'] = pd.to_datetime(objects_df.first_investment_at, format='%Y-%m-%d')
objects_df['last_investment_at'] = pd.to_datetime(objects_df.last_investment_at, format='%Y-%m-%d')
objects_df['first_funding_at'] = pd.to_datetime(objects_df.first_funding_at, format='%Y-%m-%d')
objects_df['last_funding_at'] = pd.to_datetime(objects_df.last_funding_at, format='%Y-%m-%d')
objects_df['first_milestone_at'] = pd.to_datetime(objects_df.first_milestone_at, format='%Y-%m-%d')
objects_df['last_milestone_at'] = pd.to_datetime(objects_df.last_milestone_at, format='%Y-%m-%d')

In [14]:
objects_df.dtypes

Unnamed: 0                      int64
object_id                      object
entity_type                    object
parent_id                      object
name                           object
category_code                  object
status                         object
founded_at             datetime64[ns]
closed_at              datetime64[ns]
country_code                   object
state_code                     object
city                           object
region                         object
first_investment_at    datetime64[ns]
last_investment_at     datetime64[ns]
investment_rounds               int64
invested_companies              int64
first_funding_at       datetime64[ns]
last_funding_at        datetime64[ns]
funding_rounds                  int64
funding_total_usd             float64
first_milestone_at     datetime64[ns]
last_milestone_at      datetime64[ns]
milestones                      int64
relationships                   int64
dtype: object

In [16]:
# Adding new column for collecting start dates

objects_df["start_dates"] = ""

In [17]:
# Adding a new column for collecting end dates

objects_df["end_dates"] = ""
objects_df.head()

Unnamed: 0 object_id entity_type parent_id                name  \
0           0       c:1     Company       NaN            Wetpaint   
1           1      c:10     Company       NaN             Flektor   
2           2     c:100     Company       NaN               There   
3           3   c:10000     Company       NaN             MYWEBBO   
4           4   c:10001     Company       NaN  THE Movie Streamer   

     category_code     status founded_at closed_at country_code  ...  \
0              web  operating 2005-10-17       NaT          USA  ...   
1      games_video   acquired        NaT       NaT          USA  ...   
2      games_video   acquired        NaT       NaT          USA  ...   
3  network_hosting  operating 2008-07-26       NaT          NaN  ...   
4      games_video  operating 2008-07-26       NaT          NaN  ...   

  first_funding_at last_funding_at funding_rounds funding_total_usd  \
0       2005-10-01      2008-05-19              3        39750000.0   
1              NaT             NaT              0               0.0   
2              NaT             NaT              0               0.0   
3              NaT             NaT              0               0.0   
4              NaT             NaT              0               0.0   

  first_milestone_at  last_milestone_at  milestones relationships start_dates  \
0         2010-09-05         2013-09-18           5            17               
1                NaT                NaT           0             6               
2         2003-02-01         2011-09-23           4            12               
3                NaT                NaT           0             0               
4                NaT                NaT           0             0               

   end_dates  
0             
1             
2             
3             
4             

[5 rows x 27 columns]

In [19]:
# Adding a new column for time elapsed

objects_df["time_elapsed"] = ""

In [20]:
objects_df = objects_df[['object_id',
 'entity_type',
 'parent_id',
 'name',
 'category_code',
 'status',
 'start_dates',
 'end_dates',
 'time_elapsed',                        
 'founded_at',
 'first_investment_at',
 'first_funding_at',
 'country_code',
 'state_code',
 'city',
 'region',
 'investment_rounds',
 'invested_companies',
 'funding_rounds',                                
 'closed_at',
 'last_investment_at',
 'last_funding_at',
 'funding_total_usd',
 'first_milestone_at',
 'last_milestone_at',
 'milestones',
 'relationships']]

In [21]:
objects_df.head()

object_id entity_type parent_id                name    category_code  \
0       c:1     Company       NaN            Wetpaint              web   
1      c:10     Company       NaN             Flektor      games_video   
2     c:100     Company       NaN               There      games_video   
3   c:10000     Company       NaN             MYWEBBO  network_hosting   
4   c:10001     Company       NaN  THE Movie Streamer      games_video   

      status start_dates end_dates time_elapsed founded_at  ...  \
0  operating                                    2005-10-17  ...   
1   acquired                                           NaT  ...   
2   acquired                                           NaT  ...   
3  operating                                    2008-07-26  ...   
4  operating                                    2008-07-26  ...   

  invested_companies funding_rounds closed_at last_investment_at  \
0                  0              3       NaT                NaT   
1                  0              0       NaT                NaT   
2                  0              0       NaT                NaT   
3                  0              0       NaT                NaT   
4                  0              0       NaT                NaT   

  last_funding_at funding_total_usd  first_milestone_at  last_milestone_at  \
0      2008-05-19        39750000.0          2010-09-05         2013-09-18   
1             NaT               0.0                 NaT                NaT   
2             NaT               0.0          2003-02-01         2011-09-23   
3             NaT               0.0                 NaT                NaT   
4             NaT               0.0                 NaT                NaT   

   milestones relationships  
0           5            17  
1           0             6  
2           4            12  
3           0             0  
4           0             0  

[5 rows x 27 columns]

In [22]:
# Filtering 'companies' only
companies_df = objects_df[objects_df['entity_type']=='Company']
companies_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates end_dates time_elapsed founded_at  ...  \
0       operating                                    2005-10-17  ...   
1        acquired                                           NaT  ...   
2        acquired                                           NaT  ...   
3       operating                                    2008-07-26  ...   
4       operating                                    2008-07-26  ...   
...           ...         ...       ...          ...        ...  ...   
196548  operating                                    2007-01-01  ...   
196549  operating                                    2007-11-01  ...   
196550  operating                                    1959-01-01  ...   
196551  operating                                    2008-07-01  ...   
196552  operating                                    2007-09-01  ...   

       invested_companies funding_rounds closed_at last_investment_at  \
0                       0              3       NaT                NaT   
1                       0              0       NaT                NaT   
2                       0              0       NaT                NaT   
3                       0              0       NaT                NaT   
4                       0              0       NaT                NaT   
...                   ...            ...       ...                ...   
196548                  0              0       NaT                NaT   
196549                  0              1       NaT                NaT   
196550                  0              0       NaT                NaT   
196551                  0              0       NaT                NaT   
196552                  0              1       NaT                NaT   

       last_funding_at funding_total_usd  first_milestone_at  \
0           2008-05-19        39750000.0          2010-09-05   
1                  NaT               0.0                 NaT   
2                  NaT               0.0          2003-02-01   
3                  NaT               0.0                 NaT   
4                  NaT               0.0                 NaT   
...                ...               ...                 ...   
196548             NaT               0.0          2013-08-12   
196549      2008-08-25          750000.0          2013-05-01   
196550             NaT               0.0          2012-01-01   
196551             NaT               0.0                 NaT   
196552      2008-06-01          475000.0          2006-07-01   

        last_milestone_at  milestones relationships  
0              2013-09-18           5            17  
1                     NaT           0             6  
2              2011-09-23           4            12  
3                     NaT           0             0  
4                     NaT           0             0  
...                   ...         ...           ...  
196548         2013-08-12           2             5  
196549         2013-05-01           1            14  
196550         2013-10-30           3            

##### Working on filling in gap of start and end dates

In [23]:
# Filling out start dates gaps by combining dates from three columns of 'Companies':
# funded_at, first_investment_at, first_funding_at.
# If conflicting dates are found in one record, the earliest is taken.

def mindate(a,b,c):
    return min(a,b,c)

companies_df['start_dates'] = companies_df.apply(lambda row : mindate(row['founded_at'], 
                     row['first_investment_at'], row['first_funding_at']), axis = 1)

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':


In [24]:
# Get the max date to fill "end_dates" gap
def maxdate(a,b,c):
    if a == 'NaT':
        return max(b,c)
    elif b == 'NaT':
        return max(a,c)
    else:
        return c
    
companies_df['end_dates'] = companies_df.apply(lambda row : maxdate(row['closed_at'], 
                      row['last_investment_at'], row['last_funding_at']), axis = 1)

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()


In [25]:
companies_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates  end_dates time_elapsed founded_at  ...  \
0       operating  2005-10-01 2008-05-19              2005-10-17  ...   
1        acquired         NaT        NaT                     NaT  ...   
2        acquired         NaT        NaT                     NaT  ...   
3       operating  2008-07-26        NaT              2008-07-26  ...   
4       operating  2008-07-26        NaT              2008-07-26  ...   
...           ...         ...        ...          ...        ...  ...   
196548  operating  2007-01-01        NaT              2007-01-01  ...   
196549  operating  2007-11-01 2008-08-25              2007-11-01  ...   
196550  operating  1959-01-01        NaT              1959-01-01  ...   
196551  operating  2008-07-01        NaT              2008-07-01  ...   
196552  operating  2007-09-01 2008-06-01              2007-09-01  ...   

       invested_companies funding_rounds closed_at last_investment_at  \
0                       0              3       NaT                NaT   
1                       0              0       NaT                NaT   
2                       0              0       NaT                NaT   
3                       0              0       NaT                NaT   
4                       0              0       NaT                NaT   
...                   ...            ...       ...                ...   
196548                  0              0       NaT                NaT   
196549                  0              1       NaT                NaT   
196550                  0              0       NaT                NaT   
196551                  0              0       NaT                NaT   
196552                  0              1       NaT                NaT   

       last_funding_at funding_total_usd  first_milestone_at  \
0           2008-05-19        39750000.0          2010-09-05   
1                  NaT               0.0                 NaT   
2                  NaT               0.0          2003-02-01   
3                  NaT               0.0                 NaT   
4                  NaT               0.0                 NaT   
...                ...               ...                 ...   
196548             NaT               0.0          2013-08-12   
196549      2008-08-25          750000.0          2013-05-01   
196550             NaT               0.0          2012-01-01   
196551             NaT               0.0                 NaT   
196552      2008-06-01          475000.0          2006-07-01   

        last_milestone_at  milestones relationships  
0              2013-09-18           5            17  
1                     NaT           0             6  
2              2011-09-23           4            12  
3                     NaT           0             0  
4                     NaT           0             0  
...                   ...         ...           ...  
196548         2013-08-12           2             5  
196549         2013-05-01           1            14  
196550         2013-10-30           3

In [26]:
# Dropping columns
cleaned_companies_df = companies_df.drop(companies_df[['founded_at','first_investment_at','first_funding_at','closed_at','last_investment_at','last_funding_at']], axis=1)
cleaned_companies_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates  end_dates time_elapsed country_code  ...  \
0       operating  2005-10-01 2008-05-19                       USA  ...   
1        acquired         NaT        NaT                       USA  ...   
2        acquired         NaT        NaT                       USA  ...   
3       operating  2008-07-26        NaT                       NaN  ...   
4       operating  2008-07-26        NaT                       NaN  ...   
...           ...         ...        ...          ...          ...  ...   
196548  operating  2007-01-01        NaT                       USA  ...   
196549  operating  2007-11-01 2008-08-25                       USA  ...   
196550  operating  1959-01-01        NaT                       USA  ...   
196551  operating  2008-07-01        NaT                       USA  ...   
196552  operating  2007-09-01 2008-06-01                       USA  ...   

                     city             region investment_rounds  \
0                 Seattle            Seattle                 0   
1             Culver City        Los Angeles                 0   
2               San Mateo             SF Bay                 0   
3                     NaN            unknown                 0   
4                     NaN            unknown                 0   
...                   ...                ...               ...   
196548      San Francisco             SF Bay                 0   
196549           San Jose             SF Bay                 0   
196550       Falls Church      Washington DC                 0   
196551        Los Angeles        Los Angeles                 0   
196552  Pacific Palisades  Pacific Palisades                 0   

        invested_companies  funding_rounds  funding_total_usd  \
0                        0               3         39750000.0   
1                        0               0                0.0   
2                        0               0                0.0   
3                        0               0                0.0   
4                        0               0                0.0   
...                    ...             ...                ...   
196548                   0               0                0.0   
196549                   0               1           750000.0   
196550                   0               0                0.0   
196551                   0               0                0.0   
196552                   0               1           475000.0   

        first_milestone_at last_milestone_at milestones  relationships  
0               2010-09-05        2013-09-18          5             17  
1                      NaT               NaT          0              6  
2               2003-02-01        2011-09-23          4             12  
3                      NaT               NaT          0              0  
4                      NaT               NaT          0              0  
...                    ...               ...        ...            ...  
196548          2013-08-12        2013-08-12          2     

In [27]:
# Dropping nulls in end_dates
cleaned_companies_df[cleaned_companies_df['end_dates'].notna()]

object_id entity_type parent_id                           name  \
0            c:1     Company       NaN                       Wetpaint   
13        c:1001     Company       NaN                     FriendFeed   
18       c:10014     Company       NaN                        Mobclix   
19       c:10015     Company       NaN                         Fitbit   
20      c:100155     Company       NaN                           MTPV   
...          ...         ...       ...                            ...   
196537   c:99853     Company       NaN                   ChatterBlock   
196540    c:9989     Company       NaN  Shattered Reality Interactive   
196547    c:9994     Company       NaN                         Imindi   
196549    c:9995     Company       NaN                         FohBoh   
196552    c:9998     Company       NaN                 Closet Couture   

           category_code     status start_dates  end_dates time_elapsed  \
0                    web  operating  2005-10-01 2008-05-19                
13                   web   acquired  2007-10-01 2008-02-26                
18                mobile   acquired  2008-03-01 2008-09-01                
19                health  operating  2007-10-01 2013-08-13                
20             cleantech  operating  2003-01-01 2012-01-26                
...                  ...        ...         ...        ...          ...   
196537            social  operating  2011-01-11 2011-09-06                
196540       games_video     closed  2006-06-01 2009-02-01                
196547       advertising  operating  2007-01-01 2008-05-01                
196549  public_relations  operating  2007-11-01 2008-08-25                
196552         ecommerce  operating  2007-09-01 2008-06-01                

       country_code  ...               city             region  \
0               USA  ...            Seattle            Seattle   
13              USA  ...      Mountain View             SF Bay   
18              USA  ...          Palo Alto             SF Bay   
19              USA  ...      San Francisco             SF Bay   
20              USA  ...             Austin             Austin   
...             ...  ...                ...                ...   
196537          CAN  ...           Victoria           Victoria   
196540          USA  ...      San Francisco             SF Bay   
196547          NaN  ...                NaN            unknown   
196549          USA  ...           San Jose             SF Bay   
196552          USA  ...  Pacific Palisades  Pacific Palisades   

       investment_rounds  invested_companies  funding_rounds  \
0                      0                   0               3   
13                     0                   0               1   
18                     0                   0               1   
19                     0                   0               5   
20                     0                   0               3   
...                  ...                 ...             ...   
196537                 0                   0               1   
196540                 0                   0               2   
196547                 0                   0               2   
196549                 0                   0               1   
196552                 0                   0               1   

        funding_total_usd  first_milestone_at last_milestone_at milestones  \
0              39750000.0          2010-09-05        2013-09-18          5   
13              5000000.0          2008-05-01        2012-09-13          3   
18                    0.0          1995-03-01        2012-09-18          4   
19             68069200.0                 NaT               NaT          0   
20             10125293.0          2010-01-01        2010-01-01          1   
...                   ...                 ...               ...        ...   
196537           500000.0          2011-01-11        2011-01-11          1   
196540          1100000.0          2005-11-

##### There are only 31507 end date records from the companies_df
-----

#### Looking at Product from original df for start and end dates

In [28]:
# Looking at products from the original df for end dates

end_products_df = objects_df[objects_df['entity_type']=='Product']
end_products_df

object_id entity_type parent_id  \
434913       r:1     Product       c:1   
434914      r:10     Product      c:16   
434915     r:100     Product    c:1585   
434916   r:10002     Product   c:14171   
434917   r:10003     Product   c:14173   
...          ...         ...       ...   
462646    r:9995     Product   c:14164   
462647    r:9996     Product   c:14164   
462648    r:9997     Product   c:14164   
462649    r:9998     Product   c:14164   
462650    r:9999     Product   c:14164   

                                                     name category_code  \
434913                                   Wikison Wetpaint           NaN   
434914                                 Slacker Web Player           NaN   
434915                 Vemo Enterprise Workforce Planning           NaN   
434916             dexmo.com - study and revision network           NaN   
434917                              VIrtual PBX Parachute           NaN   
...                                                   ...           ...   
462646              SiteLink, listing feed for Brokerages           NaN   
462647  EDCLink, listing feed for Economic Development...           NaN   
462648                   Cmail, broadcast email marketing           NaN   
462649                      CatylistCRM, contact database           NaN   
462650                   Catylist Custom Print/Web Design           NaN   

           status start_dates end_dates time_elapsed founded_at  ...  \
434913       live                                    2006-06-01  ...   
434914  operating                                    2007-06-05  ...   
434915       live                                    2008-07-01  ...   
434916  operating                                           NaT  ...   
434917  operating                                           NaT  ...   
...           ...         ...       ...          ...        ...  ...   
462646  operating                                           NaT  ...   
462647  operating                                           NaT  ...   
462648  operating                                           NaT  ...   
462649  operating                                           NaT  ...   
462650  operating                                           NaT  ...   

       invested_companies funding_rounds closed_at last_investment_at  \
434913                  0              0       NaT                NaT   
434914                  0              0       NaT                NaT   
434915                  0              0       NaT                NaT   
434916                  0              0       NaT                NaT   
434917                  0              0       NaT                NaT   
...                   ...            ...       ...                ...   
462646                  0              0       NaT                NaT   
462647                  0              0       NaT                NaT   
462648                  0              0       NaT                NaT   
462649                  0              0       NaT                NaT   
462650                  0              0       NaT                NaT   

       last_funding_at funding_total_usd  first_milestone_at  \
434913             NaT               0.0                 NaT   
434914             NaT               0.0                 NaT   
434915             NaT               0.0                 NaT   
434916             NaT               0.0                 NaT   
434917             NaT               0.0                 NaT   
...                ...               ...                 ...   
462646             NaT               0.0                 NaT   
462647             NaT               0.0                 NaT   
462648             NaT               0.0                 NaT   
462649             NaT               0.0                 NaT   
462650             NaT               0.0                 NaT   

        last_milestone_at  milestones relationships  
434913                NaT           

In [29]:
# Dropping null values in parent_id
cleaned_end_products_df = end_products_df[end_products_df['parent_id'].notnull()]
cleaned_end_products_df

object_id entity_type parent_id  \
434913       r:1     Product       c:1   
434914      r:10     Product      c:16   
434915     r:100     Product    c:1585   
434916   r:10002     Product   c:14171   
434917   r:10003     Product   c:14173   
...          ...         ...       ...   
462646    r:9995     Product   c:14164   
462647    r:9996     Product   c:14164   
462648    r:9997     Product   c:14164   
462649    r:9998     Product   c:14164   
462650    r:9999     Product   c:14164   

                                                     name category_code  \
434913                                   Wikison Wetpaint           NaN   
434914                                 Slacker Web Player           NaN   
434915                 Vemo Enterprise Workforce Planning           NaN   
434916             dexmo.com - study and revision network           NaN   
434917                              VIrtual PBX Parachute           NaN   
...                                                   ...           ...   
462646              SiteLink, listing feed for Brokerages           NaN   
462647  EDCLink, listing feed for Economic Development...           NaN   
462648                   Cmail, broadcast email marketing           NaN   
462649                      CatylistCRM, contact database           NaN   
462650                   Catylist Custom Print/Web Design           NaN   

           status start_dates end_dates time_elapsed founded_at  ...  \
434913       live                                    2006-06-01  ...   
434914  operating                                    2007-06-05  ...   
434915       live                                    2008-07-01  ...   
434916  operating                                           NaT  ...   
434917  operating                                           NaT  ...   
...           ...         ...       ...          ...        ...  ...   
462646  operating                                           NaT  ...   
462647  operating                                           NaT  ...   
462648  operating                                           NaT  ...   
462649  operating                                           NaT  ...   
462650  operating                                           NaT  ...   

       invested_companies funding_rounds closed_at last_investment_at  \
434913                  0              0       NaT                NaT   
434914                  0              0       NaT                NaT   
434915                  0              0       NaT                NaT   
434916                  0              0       NaT                NaT   
434917                  0              0       NaT                NaT   
...                   ...            ...       ...                ...   
462646                  0              0       NaT                NaT   
462647                  0              0       NaT                NaT   
462648                  0              0       NaT                NaT   
462649                  0              0       NaT                NaT   
462650                  0              0       NaT                NaT   

       last_funding_at funding_total_usd  first_milestone_at  \
434913             NaT               0.0                 NaT   
434914             NaT               0.0                 NaT   
434915             NaT               0.0                 NaT   
434916             NaT               0.0                 NaT   
434917             NaT               0.0                 NaT   
...                ...               ...                 ...   
462646             NaT               0.0                 NaT   
462647             NaT               0.0                 NaT   
462648             NaT               0.0                 NaT   
462649             NaT               0.0                 NaT   
462650             NaT               0.0                 NaT   

        last_milestone_at  milestones relationships  
434913                NaT           

In [30]:
cleaned_end_products_df.columns.to_list

<bound method IndexOpsMixin.tolist of Index(['object_id', 'entity_type', 'parent_id', 'name', 'category_code',
       'status', 'start_dates', 'end_dates', 'time_elapsed', 'founded_at',
       'first_investment_at', 'first_funding_at', 'country_code', 'state_code',
       'city', 'region', 'investment_rounds', 'invested_companies',
       'funding_rounds', 'closed_at', 'last_investment_at', 'last_funding_at',
       'funding_total_usd', 'first_milestone_at', 'last_milestone_at',
       'milestones', 'relationships'],
      dtype='object')>

In [31]:
cleaned_end_products_df['founded_at'].unique()

array(['2006-06-01T00:00:00.000000000', '2007-06-05T00:00:00.000000000',
       '2008-07-01T00:00:00.000000000', ...,
       '2007-03-15T00:00:00.000000000', '2008-11-27T00:00:00.000000000',
       '2007-12-30T00:00:00.000000000'], dtype='datetime64[ns]')

In [32]:
cleaned_end_products_df['first_investment_at'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [33]:
cleaned_end_products_df['first_funding_at'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [34]:
# Dropping all redundant columns

ce2_products_df = cleaned_end_products_df.drop(['entity_type',  
       'category_code', 'first_investment_at', 'first_funding_at', 'country_code', 'state_code',
       'city', 'region', 'investment_rounds', 'invested_companies',
       'funding_rounds', 'funding_total_usd', 'first_milestone_at', 'last_milestone_at',
       'milestones', 'relationships'], axis=1)

ce2_products_df

object_id parent_id                                               name  \
434913       r:1       c:1                                   Wikison Wetpaint   
434914      r:10      c:16                                 Slacker Web Player   
434915     r:100    c:1585                 Vemo Enterprise Workforce Planning   
434916   r:10002   c:14171             dexmo.com - study and revision network   
434917   r:10003   c:14173                              VIrtual PBX Parachute   
...          ...       ...                                                ...   
462646    r:9995   c:14164              SiteLink, listing feed for Brokerages   
462647    r:9996   c:14164  EDCLink, listing feed for Economic Development...   
462648    r:9997   c:14164                   Cmail, broadcast email marketing   
462649    r:9998   c:14164                      CatylistCRM, contact database   
462650    r:9999   c:14164                   Catylist Custom Print/Web Design   

           status start_dates end_dates time_elapsed founded_at closed_at  \
434913       live                                    2006-06-01       NaT   
434914  operating                                    2007-06-05       NaT   
434915       live                                    2008-07-01       NaT   
434916  operating                                           NaT       NaT   
434917  operating                                           NaT       NaT   
...           ...         ...       ...          ...        ...       ...   
462646  operating                                           NaT       NaT   
462647  operating                                           NaT       NaT   
462648  operating                                           NaT       NaT   
462649  operating                                           NaT       NaT   
462650  operating                                           NaT       NaT   

       last_investment_at last_funding_at  
434913                NaT             NaT  
434914                NaT             NaT  
434915                NaT             NaT  
434916                NaT             NaT  
434917                NaT             NaT  
...                   ...             ...  
462646                NaT             NaT  
462647                NaT             NaT  
462648                NaT             NaT  
462649                NaT             NaT  
462650                NaT             NaT  

[27715 rows x 11 columns]

In [35]:
ce2_products_df['last_investment_at'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [36]:
ce2_products_df['last_funding_at'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [37]:
# Dropping last_investment_at and last_funding_at columns

products_df_final = ce2_products_df.drop(['last_investment_at', 'last_funding_at'], axis=1)
products_df_final

object_id parent_id                                               name  \
434913       r:1       c:1                                   Wikison Wetpaint   
434914      r:10      c:16                                 Slacker Web Player   
434915     r:100    c:1585                 Vemo Enterprise Workforce Planning   
434916   r:10002   c:14171             dexmo.com - study and revision network   
434917   r:10003   c:14173                              VIrtual PBX Parachute   
...          ...       ...                                                ...   
462646    r:9995   c:14164              SiteLink, listing feed for Brokerages   
462647    r:9996   c:14164  EDCLink, listing feed for Economic Development...   
462648    r:9997   c:14164                   Cmail, broadcast email marketing   
462649    r:9998   c:14164                      CatylistCRM, contact database   
462650    r:9999   c:14164                   Catylist Custom Print/Web Design   

           status start_dates end_dates time_elapsed founded_at closed_at  
434913       live                                    2006-06-01       NaT  
434914  operating                                    2007-06-05       NaT  
434915       live                                    2008-07-01       NaT  
434916  operating                                           NaT       NaT  
434917  operating                                           NaT       NaT  
...           ...         ...       ...          ...        ...       ...  
462646  operating                                           NaT       NaT  
462647  operating                                           NaT       NaT  
462648  operating                                           NaT       NaT  
462649  operating                                           NaT       NaT  
462650  operating                                           NaT       NaT  

[27715 rows x 9 columns]

In [38]:
products_df_final = products_df_final[products_df_final['closed_at'].notna()]
products_df_final

object_id parent_id               name  status start_dates end_dates  \
435039   r:10173   c:14409        Friendzbook  closed                         
435064     r:102     c:460       Sticky Notes  closed                         
435118   r:10253    c:3347           ZuiGames  closed                         
435229   r:10393   c:14624            Bitloot  closed                         
435236     r:104     c:112               Zlio  closed                         
...          ...       ...                ...     ...         ...       ...   
461825    r:8963   c:11424        PlayCrafter  closed                         
461911    r:9055   c:12152  MacFeeder Sverige  closed                         
462096    r:9291      c:12             Twicco  closed                         
462124    r:9320   c:13193    WW.DataServices  closed                         
462194    r:9404   c:13337            Artemis  closed                         

       time_elapsed founded_at  closed_at  
435039              2007-09-21 2007-09-21  
435064              2007-08-18 2010-01-01  
435118              2008-11-14 2008-11-14  
435229              2008-11-14 2009-04-01  
435236              2006-09-10 2011-09-20  
...             ...        ...        ...  
461825              2008-07-22 1993-05-12  
461911              2008-07-01 2009-06-01  
462096              2009-11-04 2009-07-31  
462124              2006-04-01 2007-03-01  
462194              2004-01-01 2005-01-01  

[189 rows x 9 columns]

In [39]:
products_df_final_2 = products_df_final[products_df_final['founded_at'].notna()]
products_df_final_2

object_id parent_id               name  status start_dates end_dates  \
435039   r:10173   c:14409        Friendzbook  closed                         
435064     r:102     c:460       Sticky Notes  closed                         
435118   r:10253    c:3347           ZuiGames  closed                         
435229   r:10393   c:14624            Bitloot  closed                         
435236     r:104     c:112               Zlio  closed                         
...          ...       ...                ...     ...         ...       ...   
461825    r:8963   c:11424        PlayCrafter  closed                         
461911    r:9055   c:12152  MacFeeder Sverige  closed                         
462096    r:9291      c:12             Twicco  closed                         
462124    r:9320   c:13193    WW.DataServices  closed                         
462194    r:9404   c:13337            Artemis  closed                         

       time_elapsed founded_at  closed_at  
435039              2007-09-21 2007-09-21  
435064              2007-08-18 2010-01-01  
435118              2008-11-14 2008-11-14  
435229              2008-11-14 2009-04-01  
435236              2006-09-10 2011-09-20  
...             ...        ...        ...  
461825              2008-07-22 1993-05-12  
461911              2008-07-01 2009-06-01  
462096              2009-11-04 2009-07-31  
462124              2006-04-01 2007-03-01  
462194              2004-01-01 2005-01-01  

[151 rows x 9 columns]

##### Only 151 more start-end_dates pairs would be available from the products_df

In [40]:
# Final product df with start and end dates
products_df_final_2['end_dates'] = products_df_final_2.apply(lambda row : (row['closed_at']), axis = 1)
products_df_final_2

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


object_id parent_id               name  status start_dates  end_dates  \
435039   r:10173   c:14409        Friendzbook  closed             2007-09-21   
435064     r:102     c:460       Sticky Notes  closed             2010-01-01   
435118   r:10253    c:3347           ZuiGames  closed             2008-11-14   
435229   r:10393   c:14624            Bitloot  closed             2009-04-01   
435236     r:104     c:112               Zlio  closed             2011-09-20   
...          ...       ...                ...     ...         ...        ...   
461825    r:8963   c:11424        PlayCrafter  closed             1993-05-12   
461911    r:9055   c:12152  MacFeeder Sverige  closed             2009-06-01   
462096    r:9291      c:12             Twicco  closed             2009-07-31   
462124    r:9320   c:13193    WW.DataServices  closed             2007-03-01   
462194    r:9404   c:13337            Artemis  closed             2005-01-01   

       time_elapsed founded_at  closed_at  
435039              2007-09-21 2007-09-21  
435064              2007-08-18 2010-01-01  
435118              2008-11-14 2008-11-14  
435229              2008-11-14 2009-04-01  
435236              2006-09-10 2011-09-20  
...             ...        ...        ...  
461825              2008-07-22 1993-05-12  
461911              2008-07-01 2009-06-01  
462096              2009-11-04 2009-07-31  
462124              2006-04-01 2007-03-01  
462194              2004-01-01 2005-01-01  

[151 rows x 9 columns]

In [41]:
# Final product df with start and end dates
products_df_final_2['start_dates'] = products_df_final_2.apply(lambda row : (row['founded_at']), axis = 1)
products_df_final_2['end_dates'] = products_df_final_2.apply(lambda row : (row['closed_at']), axis = 1)
products_df_final_2

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


object_id parent_id               name  status start_dates  end_dates  \
435039   r:10173   c:14409        Friendzbook  closed  2007-09-21 2007-09-21   
435064     r:102     c:460       Sticky Notes  closed  2007-08-18 2010-01-01   
435118   r:10253    c:3347           ZuiGames  closed  2008-11-14 2008-11-14   
435229   r:10393   c:14624            Bitloot  closed  2008-11-14 2009-04-01   
435236     r:104     c:112               Zlio  closed  2006-09-10 2011-09-20   
...          ...       ...                ...     ...         ...        ...   
461825    r:8963   c:11424        PlayCrafter  closed  2008-07-22 1993-05-12   
461911    r:9055   c:12152  MacFeeder Sverige  closed  2008-07-01 2009-06-01   
462096    r:9291      c:12             Twicco  closed  2009-11-04 2009-07-31   
462124    r:9320   c:13193    WW.DataServices  closed  2006-04-01 2007-03-01   
462194    r:9404   c:13337            Artemis  closed  2004-01-01 2005-01-01   

       time_elapsed founded_at  closed_at  
435039              2007-09-21 2007-09-21  
435064              2007-08-18 2010-01-01  
435118              2008-11-14 2008-11-14  
435229              2008-11-14 2009-04-01  
435236              2006-09-10 2011-09-20  
...             ...        ...        ...  
461825              2008-07-22 1993-05-12  
461911              2008-07-01 2009-06-01  
462096              2009-11-04 2009-07-31  
462124              2006-04-01 2007-03-01  
462194              2004-01-01 2005-01-01  

[151 rows x 9 columns]

In [42]:
#Dropping columns fp = final product
fp_df = products_df_final_2.drop(products_df_final_2[['closed_at', 'founded_at']], axis=1)
fp_df

object_id parent_id               name  status start_dates  end_dates  \
435039   r:10173   c:14409        Friendzbook  closed  2007-09-21 2007-09-21   
435064     r:102     c:460       Sticky Notes  closed  2007-08-18 2010-01-01   
435118   r:10253    c:3347           ZuiGames  closed  2008-11-14 2008-11-14   
435229   r:10393   c:14624            Bitloot  closed  2008-11-14 2009-04-01   
435236     r:104     c:112               Zlio  closed  2006-09-10 2011-09-20   
...          ...       ...                ...     ...         ...        ...   
461825    r:8963   c:11424        PlayCrafter  closed  2008-07-22 1993-05-12   
461911    r:9055   c:12152  MacFeeder Sverige  closed  2008-07-01 2009-06-01   
462096    r:9291      c:12             Twicco  closed  2009-11-04 2009-07-31   
462124    r:9320   c:13193    WW.DataServices  closed  2006-04-01 2007-03-01   
462194    r:9404   c:13337            Artemis  closed  2004-01-01 2005-01-01   

       time_elapsed  
435039               
435064               
435118               
435229               
435236               
...             ...  
461825               
461911               
462096               
462124               
462194               

[151 rows x 7 columns]

In [121]:
merged_df = pd.merge(cleaned_companies_df, fp_df, how='left', on=['object_id'])
merged_df

object_id entity_type parent_id_x              name_x  \
0            c:1     Company         NaN            Wetpaint   
1           c:10     Company         NaN             Flektor   
2          c:100     Company         NaN               There   
3        c:10000     Company         NaN             MYWEBBO   
4        c:10001     Company         NaN  THE Movie Streamer   
...          ...         ...         ...                 ...   
196548   c:99940     Company         NaN           Webgility   
196549    c:9995     Company         NaN              FohBoh   
196550    c:9996     Company         NaN                 CSC   
196551    c:9997     Company         NaN      Top-candidate™   
196552    c:9998     Company         NaN      Closet Couture   

           category_code   status_x start_dates_x end_dates_x time_elapsed_x  \
0                    web  operating    2005-10-01  2008-05-19                  
1            games_video   acquired           NaT         NaT                  
2            games_video   acquired           NaT         NaT                  
3        network_hosting  operating    2008-07-26         NaT                  
4            games_video  operating    2008-07-26         NaT                  
...                  ...        ...           ...         ...            ...   
196548         ecommerce  operating    2007-01-01         NaT                  
196549  public_relations  operating    2007-11-01  2008-08-25                  
196550        consulting  operating    1959-01-01         NaT                  
196551            search  operating    2008-07-01         NaT                  
196552         ecommerce  operating    2007-09-01  2008-06-01                  

       country_code  ... first_milestone_at last_milestone_at milestones  \
0               USA  ...         2010-09-05        2013-09-18          5   
1               USA  ...                NaT               NaT          0   
2               USA  ...         2003-02-01        2011-09-23          4   
3               NaN  ...                NaT               NaT          0   
4               NaN  ...                NaT               NaT          0   
...             ...  ...                ...               ...        ...   
196548          USA  ...         2013-08-12        2013-08-12          2   
196549          USA  ...         2013-05-01        2013-05-01          1   
196550          USA  ...         2012-01-01        2013-10-30          3   
196551          USA  ...                NaT               NaT          0   
196552          USA  ...         2006-07-01        2008-09-10          2   

        relationships  parent_id_y  name_y  status_y start_dates_y  \
0                  17          NaN     NaN       NaN           NaT   
1                   6          NaN     NaN       NaN           NaT   
2                  12          NaN     NaN       NaN           NaT   
3                   0          NaN     NaN       NaN           NaT   
4                   0          NaN     NaN       NaN           NaT   
...               ...          ...     ...       ...           ...   
196548              5          NaN     NaN       NaN           NaT   
196549             14          NaN     NaN       NaN           NaT   
196550             44          NaN     NaN       NaN           NaT   
196551              1          NaN     NaN       NaN           NaT   
196552              1          NaN     NaN       NaN           NaT   

       end_dates_y  time_elapsed_y  
0              NaT             NaN  
1              NaT             NaN  
2              NaT             NaN  
3              NaT             NaN  
4              NaT             NaN  
...            ...             ...  
196548         NaT             NaN  
196549         NaT             NaN  
196550         NaT             NaN  
196551         NaT             NaN  
196552         NaT             NaN  

[196553 rows x 27 columns]

In [122]:
merged_df.columns.to_list

<bound method IndexOpsMixin.tolist of Index(['object_id', 'entity_type', 'parent_id_x', 'name_x', 'category_code',
       'status_x', 'start_dates_x', 'end_dates_x', 'time_elapsed_x',
       'country_code', 'state_code', 'city', 'region', 'investment_rounds',
       'invested_companies', 'funding_rounds', 'funding_total_usd',
       'first_milestone_at', 'last_milestone_at', 'milestones',
       'relationships', 'parent_id_y', 'name_y', 'status_y', 'start_dates_y',
       'end_dates_y', 'time_elapsed_y'],
      dtype='object')>

In [123]:
# checking to see if there are any not NaT values
merged_df['start_dates_y'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [124]:
# checking for any unique values
merged_df['end_dates_y'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [125]:
# Dropping all redundant columns

clean_merge_df = merged_df.drop(['parent_id_y',
       'name_y', 'status_y', 'start_dates_y', 'end_dates_y', 'time_elapsed_y'], axis=1)

clean_merge_df

object_id entity_type parent_id_x              name_x  \
0            c:1     Company         NaN            Wetpaint   
1           c:10     Company         NaN             Flektor   
2          c:100     Company         NaN               There   
3        c:10000     Company         NaN             MYWEBBO   
4        c:10001     Company         NaN  THE Movie Streamer   
...          ...         ...         ...                 ...   
196548   c:99940     Company         NaN           Webgility   
196549    c:9995     Company         NaN              FohBoh   
196550    c:9996     Company         NaN                 CSC   
196551    c:9997     Company         NaN      Top-candidate™   
196552    c:9998     Company         NaN      Closet Couture   

           category_code   status_x start_dates_x end_dates_x time_elapsed_x  \
0                    web  operating    2005-10-01  2008-05-19                  
1            games_video   acquired           NaT         NaT                  
2            games_video   acquired           NaT         NaT                  
3        network_hosting  operating    2008-07-26         NaT                  
4            games_video  operating    2008-07-26         NaT                  
...                  ...        ...           ...         ...            ...   
196548         ecommerce  operating    2007-01-01         NaT                  
196549  public_relations  operating    2007-11-01  2008-08-25                  
196550        consulting  operating    1959-01-01         NaT                  
196551            search  operating    2008-07-01         NaT                  
196552         ecommerce  operating    2007-09-01  2008-06-01                  

       country_code  ...               city             region  \
0               USA  ...            Seattle            Seattle   
1               USA  ...        Culver City        Los Angeles   
2               USA  ...          San Mateo             SF Bay   
3               NaN  ...                NaN            unknown   
4               NaN  ...                NaN            unknown   
...             ...  ...                ...                ...   
196548          USA  ...      San Francisco             SF Bay   
196549          USA  ...           San Jose             SF Bay   
196550          USA  ...       Falls Church      Washington DC   
196551          USA  ...        Los Angeles        Los Angeles   
196552          USA  ...  Pacific Palisades  Pacific Palisades   

       investment_rounds  invested_companies  funding_rounds  \
0                      0                   0               3   
1                      0                   0               0   
2                      0                   0               0   
3                      0                   0               0   
4                      0                   0               0   
...                  ...                 ...             ...   
196548                 0                   0               0   
196549                 0                   0               1   
196550                 0                   0               0   
196551                 0                   0               0   
196552                 0                   0               1   

        funding_total_usd  first_milestone_at last_milestone_at milestones  \
0              39750000.0          2010-09-05        2013-09-18          5   
1                     0.0                 NaT               NaT          0   
2                     0.0          2003-02-01        2011-09-23          4   
3                     0.0                 NaT               NaT          0   
4                     0.0                 NaT               NaT          0   
...                   ...                 ...               ...        ...   
196548                0.0          2013-08-12        2013-08-12          2   
196549           750000.0          2013-05-01        2013-05-01          1   
196550       

In [126]:
clean_merge_df = clean_merge_df.rename(columns = {'parent_id_x': 'parent_id', 
                'name_x':'name', 'status_x':'status', 'start_dates_x':'start_dates',
                'end_dates_x':'end_dates', 'time_elapsed_x':'time_elapsed_months'})

clean_merge_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates  end_dates time_elapsed_months country_code  \
0       operating  2005-10-01 2008-05-19                              USA   
1        acquired         NaT        NaT                              USA   
2        acquired         NaT        NaT                              USA   
3       operating  2008-07-26        NaT                              NaN   
4       operating  2008-07-26        NaT                              NaN   
...           ...         ...        ...                 ...          ...   
196548  operating  2007-01-01        NaT                              USA   
196549  operating  2007-11-01 2008-08-25                              USA   
196550  operating  1959-01-01        NaT                              USA   
196551  operating  2008-07-01        NaT                              USA   
196552  operating  2007-09-01 2008-06-01                              USA   

        ...               city             region investment_rounds  \
0       ...            Seattle            Seattle                 0   
1       ...        Culver City        Los Angeles                 0   
2       ...          San Mateo             SF Bay                 0   
3       ...                NaN            unknown                 0   
4       ...                NaN            unknown                 0   
...     ...                ...                ...               ...   
196548  ...      San Francisco             SF Bay                 0   
196549  ...           San Jose             SF Bay                 0   
196550  ...       Falls Church      Washington DC                 0   
196551  ...        Los Angeles        Los Angeles                 0   
196552  ...  Pacific Palisades  Pacific Palisades                 0   

        invested_companies  funding_rounds  funding_total_usd  \
0                        0               3         39750000.0   
1                        0               0                0.0   
2                        0               0                0.0   
3                        0               0                0.0   
4                        0               0                0.0   
...                    ...             ...                ...   
196548                   0               0                0.0   
196549                   0               1           750000.0   
196550                   0               0                0.0   
196551                   0               0                0.0   
196552                   0               1           475000.0   

        first_milestone_at last_milestone_at milestones  relationships  
0               2010-09-05        2013-09-18          5             17  
1                      NaT               NaT          0              6  
2               2003-02-01        2011-09-23          4             12  
3                      NaT               NaT          0              0  
4                      NaT               NaT          0              0  
...                    ...               ...     

In [127]:
clean_merge_df.dtypes

object_id                      object
entity_type                    object
parent_id                      object
name                           object
category_code                  object
status                         object
start_dates            datetime64[ns]
end_dates              datetime64[ns]
time_elapsed_months            object
country_code                   object
state_code                     object
city                           object
region                         object
investment_rounds               int64
invested_companies              int64
funding_rounds                  int64
funding_total_usd             float64
first_milestone_at     datetime64[ns]
last_milestone_at      datetime64[ns]
milestones                      int64
relationships                   int64
dtype: object

In [128]:
clean_merge_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates  end_dates time_elapsed_months country_code  \
0       operating  2005-10-01 2008-05-19                              USA   
1        acquired         NaT        NaT                              USA   
2        acquired         NaT        NaT                              USA   
3       operating  2008-07-26        NaT                              NaN   
4       operating  2008-07-26        NaT                              NaN   
...           ...         ...        ...                 ...          ...   
196548  operating  2007-01-01        NaT                              USA   
196549  operating  2007-11-01 2008-08-25                              USA   
196550  operating  1959-01-01        NaT                              USA   
196551  operating  2008-07-01        NaT                              USA   
196552  operating  2007-09-01 2008-06-01                              USA   

        ...               city             region investment_rounds  \
0       ...            Seattle            Seattle                 0   
1       ...        Culver City        Los Angeles                 0   
2       ...          San Mateo             SF Bay                 0   
3       ...                NaN            unknown                 0   
4       ...                NaN            unknown                 0   
...     ...                ...                ...               ...   
196548  ...      San Francisco             SF Bay                 0   
196549  ...           San Jose             SF Bay                 0   
196550  ...       Falls Church      Washington DC                 0   
196551  ...        Los Angeles        Los Angeles                 0   
196552  ...  Pacific Palisades  Pacific Palisades                 0   

        invested_companies  funding_rounds  funding_total_usd  \
0                        0               3         39750000.0   
1                        0               0                0.0   
2                        0               0                0.0   
3                        0               0                0.0   
4                        0               0                0.0   
...                    ...             ...                ...   
196548                   0               0                0.0   
196549                   0               1           750000.0   
196550                   0               0                0.0   
196551                   0               0                0.0   
196552                   0               1           475000.0   

        first_milestone_at last_milestone_at milestones  relationships  
0               2010-09-05        2013-09-18          5             17  
1                      NaT               NaT          0              6  
2               2003-02-01        2011-09-23          4             12  
3                      NaT               NaT          0              0  
4                      NaT               NaT          0              0  
...                    ...               ...     

In [129]:
clean_merge_df = clean_merge_df[clean_merge_df['start_dates'].notna() & clean_merge_df['end_dates'].notna()] 
clean_merge_df

object_id entity_type parent_id                           name  \
0            c:1     Company       NaN                       Wetpaint   
13        c:1001     Company       NaN                     FriendFeed   
18       c:10014     Company       NaN                        Mobclix   
19       c:10015     Company       NaN                         Fitbit   
20      c:100155     Company       NaN                           MTPV   
...          ...         ...       ...                            ...   
196537   c:99853     Company       NaN                   ChatterBlock   
196540    c:9989     Company       NaN  Shattered Reality Interactive   
196547    c:9994     Company       NaN                         Imindi   
196549    c:9995     Company       NaN                         FohBoh   
196552    c:9998     Company       NaN                 Closet Couture   

           category_code     status start_dates  end_dates  \
0                    web  operating  2005-10-01 2008-05-19   
13                   web   acquired  2007-10-01 2008-02-26   
18                mobile   acquired  2008-03-01 2008-09-01   
19                health  operating  2007-10-01 2013-08-13   
20             cleantech  operating  2003-01-01 2012-01-26   
...                  ...        ...         ...        ...   
196537            social  operating  2011-01-11 2011-09-06   
196540       games_video     closed  2006-06-01 2009-02-01   
196547       advertising  operating  2007-01-01 2008-05-01   
196549  public_relations  operating  2007-11-01 2008-08-25   
196552         ecommerce  operating  2007-09-01 2008-06-01   

       time_elapsed_months country_code  ...               city  \
0                                   USA  ...            Seattle   
13                                  USA  ...      Mountain View   
18                                  USA  ...          Palo Alto   
19                                  USA  ...      San Francisco   
20                                  USA  ...             Austin   
...                    ...          ...  ...                ...   
196537                              CAN  ...           Victoria   
196540                              USA  ...      San Francisco   
196547                              NaN  ...                NaN   
196549                              USA  ...           San Jose   
196552                              USA  ...  Pacific Palisades   

                   region investment_rounds  invested_companies  \
0                 Seattle                 0                   0   
13                 SF Bay                 0                   0   
18                 SF Bay                 0                   0   
19                 SF Bay                 0                   0   
20                 Austin                 0                   0   
...                   ...               ...                 ...   
196537           Victoria                 0                   0   
196540             SF Bay                 0                   0   
196547            unknown                 0                   0   
196549             SF Bay                 0                   0   
196552  Pacific Palisades                 0                   0   

        funding_rounds  funding_total_usd  first_milestone_at  \
0                    3         39750000.0          2010-09-05   
13                   1          5000000.0          2008-05-01   
18                   1                0.0          1995-03-01   
19                   5         68069200.0                 NaT   
20                   3         10125293.0          2010-01-01   
...                ...                ...                 ...   
196537               1           500000.0          2011-01-11   
196540               2          1100000.0          2005-11-01   
196547               2           500000.0          2009-08-27   
196549               1           750000.0          2013-05-01   
196552               1           475000.0          2006-07-0

In [131]:
# find time elapsed in months 
date1 = clean_merge_df['start_dates'].iloc[0:]
date2 = clean_merge_df['end_dates'].iloc[0:]

clean_merge_df['time_elapsed_months'] = ((date2 - date1)/np.timedelta64(1, 'M')).round()

clean_merge_df

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


object_id entity_type parent_id                           name  \
0            c:1     Company       NaN                       Wetpaint   
13        c:1001     Company       NaN                     FriendFeed   
18       c:10014     Company       NaN                        Mobclix   
19       c:10015     Company       NaN                         Fitbit   
20      c:100155     Company       NaN                           MTPV   
...          ...         ...       ...                            ...   
196537   c:99853     Company       NaN                   ChatterBlock   
196540    c:9989     Company       NaN  Shattered Reality Interactive   
196547    c:9994     Company       NaN                         Imindi   
196549    c:9995     Company       NaN                         FohBoh   
196552    c:9998     Company       NaN                 Closet Couture   

           category_code     status start_dates  end_dates  \
0                    web  operating  2005-10-01 2008-05-19   
13                   web   acquired  2007-10-01 2008-02-26   
18                mobile   acquired  2008-03-01 2008-09-01   
19                health  operating  2007-10-01 2013-08-13   
20             cleantech  operating  2003-01-01 2012-01-26   
...                  ...        ...         ...        ...   
196537            social  operating  2011-01-11 2011-09-06   
196540       games_video     closed  2006-06-01 2009-02-01   
196547       advertising  operating  2007-01-01 2008-05-01   
196549  public_relations  operating  2007-11-01 2008-08-25   
196552         ecommerce  operating  2007-09-01 2008-06-01   

        time_elapsed_months country_code  ...               city  \
0                      32.0          USA  ...            Seattle   
13                      5.0          USA  ...      Mountain View   
18                      6.0          USA  ...          Palo Alto   
19                     70.0          USA  ...      San Francisco   
20                    109.0          USA  ...             Austin   
...                     ...          ...  ...                ...   
196537                  8.0          CAN  ...           Victoria   
196540                 32.0          USA  ...      San Francisco   
196547                 16.0          NaN  ...                NaN   
196549                 10.0          USA  ...           San Jose   
196552                  9.0          USA  ...  Pacific Palisades   

                   region investment_rounds  invested_companies  \
0                 Seattle                 0                   0   
13                 SF Bay                 0                   0   
18                 SF Bay                 0                   0   
19                 SF Bay                 0                   0   
20                 Austin                 0                   0   
...                   ...               ...                 ...   
196537           Victoria                 0                   0   
196540             SF Bay                 0                   0   
196547            unknown                 0                   0   
196549             SF Bay                 0                   0   
196552  Pacific Palisades                 0                   0   

        funding_rounds  funding_total_usd  first_milestone_at  \
0                    3         39750000.0          2010-09-05   
13                   1          5000000.0          2008-05-01   
18                   1                0.0          1995-03-01   
19                   5         68069200.0                 NaT   
20                   3         10125293.0          2010-01-01   
...                ...                ...                 ...   
196537               1           500000.0          2011-01-11   
196540               2          1100000.0          2005-11-01   
196547               2           500000.0          2009-08-27   
196549               1           750000.0          2013-05-01   
196552               1           475000.0       

In [153]:
path2 = '../FWIP_Venture/'

In [159]:
time_elapsed_clean = clean_merge_df

time_elapsed_clean.to_csv(path2 + "time_elapsed_clean.csv", index=False)

In [162]:
filename = 'time_elapsed_clean.csv'
bucket = 'team3-final-bucket'

s3.meta.client.upload_file(Filename = filename, Bucket= bucket, Key = filename)